### ISSUE TITLE
Duplicated Add Braces To Arrow Function

### ISSUE THREAD

**Issue by hediet** [p42ai/refactor-vscode#20]
![image](https://user-images.githubusercontent.com/2931520/170969090-489ee573-3d17-40a5-8e53-2a5b61c7b455.png)

One is from p42, the other from TypeScript

**Comment by lgrammel** [comment]
There is some overlap between the refactorings that P42 offers and the refactorings that are built into TypeScript.

What UX would you ideally expect here?

**Comment by hediet** [comment]
You can implement a typescript language service plugin to disable the TypeScript refactoring.

I would expect to not suggest the same refactoring twice.

**Comment by lgrammel** [comment]
I agree, seeing the same refactoring twice is not great. I'm a bit hesitant to disable functionality provided by TS or to affect their language service from the P42 extension (which can cause stability issues, especially when they upgrade to different versions). 

Maybe I'll implement it as a separate open source companion extension where you can selectively disable some TypeScript refactorings. Would this be an option for you?

**Comment by hediet** [comment]
I would ship it with the same extension. You can still add settings to disable it.

The default experience should be a good one, seeing the same refactoring twice is not good.

If you can't or don't want to disable the TypeScript "Add braces to arrow function", you could disable P42's refactoring "Add {...} to arrow function" instead.

**Comment by lgrammel** [comment]
I'll think about how to best do this, maybe there are other options as well. Disabling the P42 refactoring is safer, it might be a preferrable default option.

**Comment by lgrammel** [comment]
Fixed in `v1.112.3` for add / remove braces to arrow functions. The refactoring from the typescript language service is now hidden, except on web (because typescript does not support plugins on web yet). Thanks for the bug report!

**Comment by OliverJAsh** [comment]
> The refactoring from the typescript language service is now hidden

Is there a way to disable this? I have the following keyboard shortcuts setup to use these built-in refactorings and they no longer work:

```json
    {
        "key": "shift+alt+r",
        "command": "editor.action.codeAction",
        "args": { "kind": "refactor.rewrite.arrow.braces.remove" }
    },
    {
        "key": "shift+alt+a",
        "command": "editor.action.codeAction",
        "args": { "kind": "refactor.rewrite.arrow.braces.add" }
    },
```

<img width="626" alt="image" src="https://user-images.githubusercontent.com/921609/173151008-add21692-76cd-4e43-99c2-b89bc726131f.png">

Alternatively, is it possible to update my keyboard shortcuts to use the refactorings provided by this extension?

**Comment by lgrammel** [comment]
@OliverJAsh thanks for the bug report! Right now, both options are unfortunately not possible. I'll look into adding a setting for disabling the ts code action hiding, and also expose each refactoring with a specific kind, so it can be mapped to a shortcut.

**Comment by lgrammel** [comment]
@OliverJAsh I've just released `v1.115.1`. Starting with that version, you can assign keyboard shortcuts to the matching code actions from the JavaScript Assistant:
```
  {
    "key": "shift+alt+r",
    "command": "editor.action.codeAction",
    "args": { "kind": "refactor.convert.p42.remove-braces-from-arrow-function" }
  },
  {
    "key": "shift+alt+a",
    "command": "editor.action.codeAction",
    "args": { "kind": "refactor.convert.p42.add-braces-to-arrow-function" }
  },
```

Adding a setting to enable/disable the overlapping typescript refactorings and making this an opt-in experience (because it can affect the current setup) will come in a future release.

Thanks for reporting the issue!

**Comment by lgrammel** [comment]
@OliverJAsh I've released `v1.118.0` which allows you to choose what code assists are shown when there is an overlap:

![image](https://user-images.githubusercontent.com/205036/174055412-b8c205f4-2219-4ca7-8b65-dbee8a827495.png)

I've chosen the opt-out solution, i.e. P42 deactivates overlapping TS code actions by default, to provide a good user experience out of the box for most users.
